control-plane-operator/controllers/hostedcontrolplane/etcd/reconcile.go

```diff
@@ -1,209 +1,367 @@
-				Name: "MY_POD_NAMESPACE",
-				ValueFrom: &corev1.EnvVarSource{
-					FieldRef: &corev1.ObjectFieldSelector{
-						FieldPath: "metadata.namespace",
-					},
+func reconcileEtcdContainer(ss *appsv1.StatefulSet, c *corev1.Container, image string, replicas int) {
+	c.Image = image
+	c.ImagePullPolicy = corev1.PullAlways
+	c.Command = []string{"/usr/bin/etcd-cloud-operator", "--log-level", "error", "--config", "/etc/etcd/config/config.yaml"}
+	c.VolumeMounts = []corev1.VolumeMount{
+		{
+			Name:      "data",
+			MountPath: "/var/lib",
+		},
+		{
+			Name:      "config",
+			MountPath: "/etc/etcd/config",
+		},
+		{
+			Name:      "peer-tls",
+			MountPath: "/etc/etcd/tls/peer",
+		},
+		{
+			Name:      "server-tls",
+			MountPath: "/etc/etcd/tls/server",
+		},
+	}
+	c.Env = []corev1.EnvVar{
+		{
+			Name:  "ETCD_API",
+			Value: "3",
+		},
+		{
+			Name:  "ETCDCTL_INSECURE_SKIP_TLS_VERIFY",
+			Value: "true",
+		},
+		{
+			Name:  "STATEFULSET_SERVICE_NAME",
+			Value: manifests.EtcdDiscoveryService(ss.Namespace).Name,
+		},
+		{
+			Name:  "STATEFULSET_NAME",
+			Value: ss.Name,
+		},
+		{
+			Name:  "STATEFULSET_DNS_CLUSTER_SUFFIX",
+			Value: "cluster.local",
+		},
+		{
+			Name: "POD_IP",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "status.podIP",
 				},
 			},
-			{
-				Name: "MY_POD_NAME",
-				ValueFrom: &corev1.EnvVarSource{
-					FieldRef: &corev1.ObjectFieldSelector{
-						FieldPath: "metadata.name",
-					},
+		},
+		{
+			Name: "STATEFULSET_NAMESPACE",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.namespace",
 				},
 			},
-		}
+		},
+		// TODO: Find a way to avoid encoding this in env
+		{
+			Name:  "STATEFULSET_REPLICAS",
+			Value: strconv.Itoa(replicas),
+		},
+	}
+	c.Ports = []corev1.ContainerPort{
+		{
+			Name:          "client",
+			ContainerPort: 2379,
+			Protocol:      corev1.ProtocolTCP,
+		},
+		{
+			Name:          "http",
+			ContainerPort: 2378,
+			Protocol:      corev1.ProtocolTCP,
+		},
+		{
+			Name:          "peer",
+			ContainerPort: 2380,
+			Protocol:      corev1.ProtocolTCP,
+		},
+		{
+			Name:          "metrics",
+			ContainerPort: 2381,
+			Protocol:      corev1.ProtocolTCP,
+		},
+	}
+	// TODO: Extract probes into DeploymentConfig
+	c.LivenessProbe = &corev1.Probe{
+		Handler: corev1.Handler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: "/status",
+				Port: intstr.Parse("http"),
+			},
+		},
+		InitialDelaySeconds: 1,
+		PeriodSeconds:       5,
+		FailureThreshold:    6,
+	}
+	c.ReadinessProbe = &corev1.Probe{
+		Handler: corev1.Handler{
+			Exec: &corev1.ExecAction{
+				Command: []string{"/bin/sh", "-c",
+					"/usr/bin/etcdctl --cacert /etc/etcd/tls/server/server-ca.crt --cert /etc/etcd/tls/server/server.crt --key /etc/etcd/tls/server/server.key --endpoints=${HOSTNAME}:2379 endpoint health"},
+			},
+		},
+		// TODO: The client port is mTLS secured, so TCPSocket won't work
```

Comment:
Is this a real `TODO`? The `exec etcdctl` is also used by OCP itself so this is unlikely to change anytime soon...?